### PR TITLE
Add stop loss timeout and add error handling for getting actual profit

### PR DESCRIPTION
**Pull request by spggooss**
This PR adds the stop loss time out, it also adds error handling for getting the actual profit. Sometimes this could result in an empty string when the deal on 3commas wasn't valid any more. A reason for this could be that the coin was delete from the exchange, or changed its name on the exchange.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -464,6 +464,9 @@ If the value is above or equal to activation-percentage, the initial SL is calcu
 
 `new_stoploss = initial-stoploss + (actual_profit_percentage - activation_percentage)`
 
+The script also supports a timeout for the stop loss. This means that it will add a configured timeout for the stop loss to be activated. When the timeout is 60 seconds for example, when the price reaches the stop loss price, a counter on 3commas will run to count down to 0.
```
Comment (amargedon) [inline README.md]: This script also supports a configurable timeout for the stoploss, which will be activated (and updated) when the stoploss is activated. For example, with a timeout of 60 seconds, 3Commas will sell the deal (market order) when after 60 seconds the current price is still below the set stoploss price. 

Misschien wat compacter, en voor mijn gevoel daardoor duidelijker?

File: README.md [hunk 2]
```diff
@@ -503,23 +506,25 @@ Author of this script is [amargedon](https://github.com/amargedon).
...
+- **3c-apisecret** - your 3Commas API key secret value.
+- **notifications** - set to true to enable notifications. (default = False)
+- **notify-urls** - one or a list of apprise notify urls, each in " " seperated with commas. See [Apprise website](https://github.com/caronc/apprise) for more information.
+- *[tsl_tp_]*
+- **botids** - a list of bot id's to manage separated with commas.
+- **config** - a list of objects with the settings for that percentage.
+- *object*
+- **activation-percentage** - from % of profit this object is valid for.
+- **initial-stoploss-percentage** - % of stoploss to set when activation-percentage is reached.
+- **sl-timeout** - timeout before stoploss is triggered. 3Commas still handles triggering of the stop loss, but the market buy order is only triggerd after the timeout has expired and the current price is still under the stop loss price. ***(Be careful, the market order will also be activated when the profit is below 0%. This means the deal could end up selling at a loss!)***  (Default is 0 which means it is turned off).
```
Comment (amargedon) [inline README.md]: Dit staat boven in de tekst al uitgelegd, dus ik zou dit niet twee keer erin willen hebben. Dit lijstje is ook als kort overzicht bedoeld, dus zou eigenlijk iets als 'stoploss timeout in seconds. (default = 0)' verwachten.

File: README.md [hunk 3]
```diff
@@ -536,7 +541,7 @@ notify-urls = [ "tgram://9995888120:BoJPor6opeHyxx5VVZPX-BoJPor6opeHyxx5VVZPX/"
 
 [tsl_tp_default]
 botids = [ 123456 ]
-config = [{"activation-percentage": "2.0","initial-stoploss-percentage": "0.5","sl-increment-factor": "0.0","tp-increment-factor": "0.0"},{"activation-percentage": "3.0","initial-stoploss-percentage": "2.0","sl-increment-factor": "0.4","tp-increment-factor": "0.4"}]
+config = [{"activation-percentage": "2.0","initial-stoploss-percentage": "0.5","sl-timeout": "600","sl-increment-factor": "0.0","tp-increment-factor": "0.0"},{"activation-percentage": "3.0","initial-stoploss-percentage": "2.0","sl-timeout": "800","sl-increment-factor": "0.4","tp-increment-factor": "0.4"}]
```
Comment (amargedon) [inline README.md]: Ik zou een van beiden 0 laten, om duidelijk te laten zien dat dat ook kan/werkt plus dat je dan beide situaties hebt; aan en uit.

File: trailingstoploss_tp.py [hunk 4]
```diff
@@ -108,39 +110,67 @@ def upgrade_config(thelogger, cfg):
...
+                    }
+                    jsonconfiglist.append(cfgsectionconfig)
+
+            if len(jsonconfiglist) > 0:
+                cfg.set(cfgsection, "config", json.dumps(jsonconfiglist))
+
+                with open(f"{datadir}/{program}.ini", "w+") as cfgfile:
+                    cfg.write(cfgfile)
+
+                thelogger.info("Upgraded section %s to have config list" % cfgsection)
```
Comment (amargedon) [inline trailingstoploss_tp.py]: Netjes gedaan! Zou deze melding wel veranderen naar iets als 'Updates section %s to add sl-timeout'.

File: trailingstoploss_tp.py [hunk 5]
```diff
@@ -108,39 +110,67 @@ def upgrade_config(thelogger, cfg):
...
-        },
+        payload=payload,
     )
     if data:
         logger.info(
             f"Changing SL for deal {deal['pair']} ({deal_id}) on bot \"{bot_name}\"\n"
             f"Changed SL from {deal['stop_loss_percentage']}% to {new_stoploss}%. "
-            f"Changed TP from {deal['take_profit']}% to {new_take_profit}%"
+            f"Changed TP from {deal['take_profit']}% to {new_take_profit}% "
+            f"Changed SL timeout to {sl_timeout}s."
```
Comment (amargedon) [inline trailingstoploss_tp.py]: Is het niet handig om voor de SL timeout ook een from - to te laten zien? Blijft het consistent met de rest namelijk, ookal zou het kunnen dat de waarden hetzelfde zijn gebleven.

File: trailingstoploss_tp.py [hunk 6]
```diff
@@ -108,39 +110,67 @@ def upgrade_config(thelogger, cfg):
...
+        "stop_loss_percentage": new_stoploss,
+        "take_profit": new_take_profit,
+    }
+
+    # Check whether there is a timeout set for this config, when there is, set the timeout to this value.
+    # Otherwise, send the payload as it is.
+    if sl_timeout != 0:
+        payload["stop_loss_timeout_enabled"] = True
+        payload["stop_loss_timeout_in_seconds"] = sl_timeout
+
```
Comment (amargedon) [inline trailingstoploss_tp.py]: Dit stukje gaat, denk ik, niet helemaal goed werken. Volgens mij moeten we altijd de sl-timeout overnemen die binnenkomt en die moet ook 0 kunnen zijn. Stel je eens het volgende voor; in de config heb ik twee profit configs. In de eerste heb ik een sl-timeout van 600 omdat ik verwacht dat de deal wat meer zal bouncen, in de tweede wil ik de sl-timeout op 0 hebben want dan wil ik gewoon sluiten. 

Zoals de code nu is, zal de 0 uit de tweede config nooit meer doorkomen en mijn sl-timeout dus altijd op 600 blijven staan in de deal.

File: trailingstoploss_tp.py [hunk 7]
```diff
@@ -213,29 +243,37 @@ def process_deals(thebot, section_config):
             deal_id = deal["id"]
 
             if deal["strategy"] in ("short", "long"):
-                current_deals.append(deal_id)
+                # Check whether the actual_profit_percentage can be obtained from the deal,
+                # If it can't, skip this deal.
+                try:
+                    current_deals.append(deal_id)
```
Comment (amargedon) [inline trailingstoploss_tp.py]: Persoonlijk vind ik een try/catch niet zo heel mooi hiervoor, des te meer omdat er ook een andere value ongeldig kan zijn en in de catch rapporteren we dan dat het altijd de profit is die ongeldig is...maar dat hoeft dus niet!

Ik had zelf in gedachten om geen try/catch te gebruiken, maar waar je nu de try hebt staan een if toe te voegen:

if deal["actual_profit_percentage"] is "":
   <logmelding met dat de actual_profit_percentage voor deal x ongeldig is>
   continue;
   
Dan is specifiek duidelijk op welke fout we controleren, en wat er dan fout is. Andere checks kunnen dan later op dezelfde manier toegevoegd worden.

File: trailingstoploss_tp.py [hunk 8]
```diff
@@ -318,6 +356,8 @@ def handle_new_deal(thebot, deal, profit_config):
 
     activation_percentage = float(profit_config.get("activation-percentage"))
 
+    sl_timeout = float(profit_config.get("sl-timeout"))
```
Comment (amargedon) [inline trailingstoploss_tp.py]: Waarom hier geen cast naar int, zodat je dat onderin bij de logregel niet meer hoeft te doen? Nu is het twee keer een cast, beetje zonde.

File: trailingstoploss_tp.py [hunk 9]
```diff
@@ -430,6 +477,16 @@ def handle_update_deal(thebot, deal, existing_deal, profit_config):
...
+                # Check whether there is an old timeout, and log when the new time out is higher than the old time out
+                old_sl_timeout = deal["stop_loss_timeout_in_seconds"]
+                if old_sl_timeout is not None:
+                    if sl_timeout > old_sl_timeout:
+                        logger.info(
+                            f"StopLoss timeout increased from {old_sl_timeout}s "
+                            f"to {int(sl_timeout)}s",
+                            True
+                        )
+
```
Comment (amargedon) [inline trailingstoploss_tp.py]: Waarom mag de timeout alleen maar groter worden? Zie mijn eerdere comment; ik kan best een timeout van 100, 50, 250, en 0 en vier opeenvolgende configs willen gebruiken. Dus volgens mij zou de check hier moeten zijn dat de oude en nieuwe ongelijk zijn aan elkaar. Overigens kan je overwegen om beide if statements in een and samen te voegen, maakt het iets compacter.

Qua naamgeving heb ik overal 'current' en 'new' gebruikt, zou mooi zijn om in die lijn te blijven.

**Review by amargedon — CHANGES_REQUESTED** [review]
Overall gezien een goede toevoeging, en heel goed hoe je dat zo snel in de structuur hebt weten toe te voegen (het is niet een heel gemakkelijk script). Wel denk ik dat de afhandeling van de sl-timeout bepaalde use-cases niet toelaat en dat moet wel anders.

Verder een aantal schoonheidszaken, wat ook een persoonlijke voorkeur kan zijn van mijn kant kan zijn. Zie het niet als afbreuk op je werk, want nogmaals, wat je hebt gedaan is goed werk!

**Review by amargedon — APPROVED** [review]
